### PR TITLE
made ckd_priv function private and impl AsRef for ChildNumber

### LINKED ISSUE #1866: `ExtendedPrivKey::ckd_priv` is unreadable
Apart from random-looking name it's not immediately obvious how does it differ from `derive_priv`. We should just make it private and possibly inline it into the for loop.

If a special method for single-child derivation is still desired (despite `&[child_numer]` being viable) we could just `impl AsRef<[ChildNumber]> for ChildNumber` so that `derive_priv` would work.

### PR DESCRIPTION
**Pull request by startup-dreamer**
closes #1866

### CODE REVIEW

**Comment by apoelstra** [comment]
I think we should update the docs for `derive_priv` to say that you can now pass a lone `ChildNumber` as the path, and we should add a unit test for this as well.

File: bitcoin/src/bip32.rs [hunk 1]
```diff
@@ -209,6 +209,10 @@ impl FromStr for ChildNumber {
     }
 }
 
+impl AsRef<[ChildNumber] for ChildNumber{
```
Comment (junderw) [inline bitcoin/src/bip32.rs:212]: Missing closing >

File: bitcoin/src/bip32.rs [hunk 2]
```diff
@@ -209,6 +209,10 @@ impl FromStr for ChildNumber {
...
+    fn as_ref(&self) -> &[ChildNumber] { &self.0 }
```
Comment (junderw) [inline bitcoin/src/bip32.rs:213]: [`core::ptr::slice_from_raw_parts`](https://doc.rust-lang.org/stable/core/ptr/fn.slice_from_raw_parts.html) should be used to convert the value into a 1 size slice (requires unsafe).

Unless there's another way to do this I'm missing...

File: bitcoin/src/bip32.rs [hunk 3]
```diff
@@ -557,13 +561,13 @@ impl ExtendedPrivKey {
...
         for cnum in path.as_ref() {
-            sk = sk.ckd_priv(secp, *cnum)?;
+            sk = sk.ckd_priv_single(secp, *cnum)?;
         }
         Ok(sk)
     }
 
     /// Private->Private child key derivation
-    pub fn ckd_priv<C: secp256k1::Signing>(
+    fn ckd_priv_single<C: secp256k1::Signing>(
```
Comment (junderw) [inline bitcoin/src/bip32.rs:570]: name change isn't necessary. Just remove the pub.

**Review by junderw — CHANGES_REQUESTED** [review]
This doesn't compile, so I suggested some changes.

File: bitcoin/src/bip32.rs (hunk 2 above)
Comment (Kixunil) [reply bitcoin/src/bip32.rs:213]: [`core::slice::from_ref`](https://doc.rust-lang.org/stable/core/slice/fn.from_ref.html)

**Comment by startup-dreamer** [comment]
Just curious why CI nightly toolchain failing, Is there anything which will fix this

**Comment by tcharding** [comment]
Just check the logs, it shows the last bash command to fail
```
+ cargo fmt --check
```

**Review by junderw — CHANGES_REQUESTED** [review]
Should use [`core::slice::from_ref`](https://doc.rust-lang.org/stable/core/slice/fn.from_ref.html) as @Kixunil suggested.

**Comment by startup-dreamer** [comment]
> Should use [`core::slice::from_ref`](https://doc.rust-lang.org/stable/core/slice/fn.from_ref.html) as @Kixunil suggested.

Done

**Comment by apoelstra** [comment]
Can you squash these commits?

**Comment by startup-dreamer** [comment]
@apoelstra commits reverted while squashing should we close this PR and i will make a new PR with all the required changes

**Comment by junderw** [comment]
... That's not how squashing works.........

Squashing 3 commits means that you have condensed all the changes of 3 commits into 1 commit. The end state should be the same.

**Comment by startup-dreamer** [comment]
> ... That's not how squashing works.........
> 
> Squashing 3 commits means that you have condensed all the changes of 3 commits into 1 commit. The end state should be the same.

yes, I did something wrong while squashing i guess it should be fine now, Thank You everyone for your help and guidance i am still learning and want to contribute.

**Comment by apoelstra** [comment]
Oops, can you also run `cargo +nighty fmt` and amend the commit?

**Review by apoelstra — DISMISSED** [review]
ACK ffe7b057f44c71ed1e8883ef2616bc46b70b413c

**Review by Kixunil — COMMENTED** [review]
Apart from formatting this looks OK.

**Comment by startup-dreamer** [comment]
extremely sorry for late reply should be able to merge now

**Review by Kixunil — APPROVED** [review]
ACK 4be019180b279b1291fc1277555fed7821b12362

**Review by junderw — APPROVED** [review]
ACK 4be019180b279b1291fc1277555fed7821b12362

**Review by apoelstra — APPROVED** [review]
ACK 4be019180b279b1291fc1277555fed7821b12362

**Pull request merged**